export/kafka.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	return agent
+}
+
+func createTopic(topicInfo TopicConfig, kafkaConfig *kafka.ConfigMap) {
+	adminClient, err := kafka.NewAdminClient(kafkaConfig)
+	if err != nil {
+		fmt.Printf("Failed to create Admin client: %s\n", err)
+		os.Exit(1)
+	}
+
+	// Contexts are used to abort or limit the amount of time
+	// the Admin call blocks waiting for a result.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create topics on cluster.
+	// Set Admin options to wait for the operation to finish (or at most 60s)
+	maxDuration, err := time.ParseDuration("60s")
+	if err != nil {
+		panic("time.ParseDuration(60s)")
+	}
+
+	if topicInfo.NumPartitions != 0 && topicInfo.NumReplicas != 0 {
+		results, err := adminClient.CreateTopics(ctx,
+			[]kafka.TopicSpecification{{
+				Topic:             topicInfo.Topic,
+				NumPartitions:     topicInfo.NumPartitions,
+				ReplicationFactor: topicInfo.NumReplicas,
+			}},
+			kafka.SetAdminOperationTimeout(maxDuration))
+
+		if err != nil {
+			fmt.Printf("Problem during the topic creation: %v\n", err)
+			os.Exit(1)
+		}
+
+		// Check for specific topic errors
+		for _, result := range results {
+			if result.Error.Code() != kafka.ErrNoError &&
+				result.Error.Code() != kafka.ErrTopicAlreadyExists {
+				fmt.Printf("Topic creation failed for %s: %v",
+					result.Topic, result.Error.String())
+				os.Exit(1)
+			}
+		}
+	}
+
+	adminClient.Close()
+
+}
+
+// Stop closes the Kafka producer.
+func (e Kafka) Stop() {
+	defer e.producer.Close()
+}
+
+// SetMessageFlushTime sets the time to wait to flush the
+// Kafka message buffer. Default is 15 seconds
+func (e *ExporterAgent) SetMessageFlushTime(seconds int) {
+	newKafka := e.Exporter.(Kafka)
+	newKafka.messageFlushTimeSec = seconds
+	e.Exporter = newKafka
+}
+
+// ExportMetrics converts the metrics to a metrics service request protobuf and
+// makes a POST request with that payload to a Kafka broker.
+func (e Kafka) ExportMetrics(ctx context.Context, data []*metricdata.Metric) error {
+	go handleEvents(e.producer.Events())
+
+	for _, d := range data {
+		if matched, _ := regexp.Match(e.config.IncludeFilter, []byte(d.Descriptor.Name)); matched {
+			d.Resource, _ = TotDetector(ctx)
+			metricsRequestpb := metricToProto(d)
+			payload, err := proto.Marshal(metricsRequestpb)
+			if err != nil {
+				log.Printf("Marshalling Error: %v", err)
+			}
+
+			err = e.producer.Produce(&kafka.Message{
+				TopicPartition: kafka.TopicPartition{
+					Topic:     &e.topicInfo.Topic,
+					Partition: kafka.PartitionAny,
+				},
+				Value: payload,
+			}, nil)
+
+			if err != nil {
+				log.Printf("Error sending message with Producer: %v", err)
+			}
+		}
+	}
+
+	droppedCount := e.producer.Flush(e.messageFlushTimeSec * 1000)
+	stats.Record(context.Background(), messagesDropped.M(int64(droppedCount)))
+	droppedDelta := droppedCount - e.lastDroppedLogCount
+
+	if droppedDelta > 0 {
+		log.Println("Failed to produce %i metrics messages", droppedDelta)
```

Comment:
do not log within the library. You may return the droppedDelta as a return parameter if it is important for the library caller.

Reply:
I did this in response to one of @jcrobak comments. It was to rate limit the error messages in a similar manner as the Java reporter: https://github.com/confluentinc/ce-kafka/blob/f24c695758a2628759659f22cfd8efd387f943c4/ce-metrics/src/main/java/io/confluent/telemetry/exporter/kafka/KafkaExporter.java#L165-L173

Do you think it would be a better idea for the exporter to just keep track of droppedDelta and if the user wants, they can do their own check of droppedDelta?

Reply:
> I did this in response to one of @jcrobak comments. It was to rate limit the error messages in a similar manner as the Java reporter: https://github.com/confluentinc/ce-kafka/blob/f24c695758a2628759659f22cfd8efd387f943c4/ce-metrics/src/main/java/io/confluent/telemetry/exporter/kafka/KafkaExporter.java#L165-L173
> 
> Do you think it would be a better idea for the exporter to just keep track of droppedDelta and if the user wants, they can do their own check of droppedDelta?

Yes.  golang logger is not as standard as in java. It is better to have main app to control how to log the error properly.